src/Components/CreateImageWizard/formComponents/AWSSourcesSelect.js

```diff
@@ -55,7 +55,7 @@ export const AWSSourcesSelect = ({
   useEffect(() => {
     if (isFetchingDetails || !isSuccessDetails) return;
     change('aws-associated-account-id', sourceDetails?.aws?.account_id);
-  }, [isFetchingDetails, isSuccessDetails]);
+  });
```

Comment:
I tried to change this part of the code to   const [isFetch, setIsFetch] = useState(isFetchingDetails);
  const [isFinish, SetIsFinish] = useState(isSuccessDetails);
  if (isFetchingDetails || !isSuccessDetails) return;
  if (isFetchingDetails !== isFetch || isSuccessDetails !== isFinish) {
    setIsFetch(isFetchingDetails);
    SetIsFinish(isSuccessDetails);
    change('aws-associated-account-id', sourceDetails?.aws?.account_id);
  }
  but its look like a lot of tests fail. 
  can you tell me if this is the direction I should go?
  @lavocatt

Reply:
I tried this solution useEffect(() => {
    if (isFetchingDetails || !isSuccessDetails) return;
    change('aws-associated-account-id', sourceDetails?.aws?.account_id);
  }, [
    change,
    sourceDetails?.aws?.account_id,
    isFetchingDetails,
    isSuccessDetails,
  ]);
  and its look like its working good

Reply:
It's not that this code chunk won't work, it's more that it is addressing the issue incorrectly.

We have to ask ourselves, why is the linter producing a warning when not all the dependencies of a `useEffect` are listed?

The reason why the react devs have produced these warnings is to alert us on bad patterns. They have a list of situations where people tend to use `useEffects` to solve problems where in reality they shouldn't. And we're on one of these cases

**First let's break down what's happening here:**

```
  useEffect(() => {
    if (isFetchingDetails || !isSuccessDetails) return;
    change('aws-associated-account-id', sourceDetails?.aws?.account_id);
  }, [isFetchingDetails, isSuccessDetails]);
  });
```

What this piece of code does: it will execute the function whenever the value `isFetchingDetails` or  `isSuccessDetails` are getting updated.

```
() => {
    if (isFetchingDetails || !isSuccessDetails) return;
    change('aws-associated-account-id', sourceDetails?.aws?.account_id);
  }
```

The function itself updates the value of the form state `aws-associated-account-id` with what's inside `sourceDetails?.aws?.account_id`. And that's it.

**Now, if we change the `useEffect` to the solution you were proposing:**

```
  useEffect(() => {
    if (isFetchingDetails || !isSuccessDetails) return;
    change('aws-associated-account-id', sourceDetails?.aws?.account_id);
  }, []);
  });
```

What's happening there, is that after every render of the component an extra render will get performed to call the inner function and update the form value `aws-associated-account-id` with the content of `sourceDetails?.aws?.account_id`.

So why the two approaches are working:
* In the first scenario the function is called only in certain conditions, but these conditions are the one we are interested in, so no issues
* In the second scenario the function is always called, which means that it necesserarily includes the conditions we are inserted in

**But what about the efficiency impact of the solution you are proposing?**

Clearly the solution before was less intrusive because it is called in less cases than every time.

Granted that we shouldn't seek performance optimizations too early in the development process, but this is an example of a performance regression. Not a bad one, but we can do better.


**So what should we do about it?**

There's a couple of possibilities for us:
* Not using a `useEffect`, there's a pattern to use instead for this https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes
* Silence the warning with an eslint comment
* Do nothing

**first solution: Not using the `useEffect`:**
```
  const [previsFetchingDetails, setPrevisFetchingDetails] = useState(isFetchingDetails);
  const [previsSuccessDetails, setPrevisSuccessDetails] = useState(isSuccessDetails);
  if (isFetchingDetails !== previsFetchingDetails || isSuccessDetails !== previsSuccessDetails) {
    setPrevisFetchingDetails(isFetchingDetails);
    setPrevisSuccessDetails(isSuccessDetails);
    if (!(isFetchingDetails || !isSuccessDetails)) {
      change('aws-associated-account-id', sourceDetails?.aws?.account_id);
    }
  }
```

I quickly wrote this by following the link https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes

So this will behave exactly like a `useEffect` except now we're compliant with the react guidelines.

**second solution: silence the warning with an esling comment:**

So the previous solution is bringing a new way of doing things in the code base, and we need a team discussion about if that's what we want to do or not. Because guidelines are one thing but what the team wants is another.
A simpler solution to address the issue would be to simply silence the warning:

```
  useEffect(() => {
    if (isFetchingDetails || !isSuccessDetails) return;
    change('aws-associated-account-id', sourceDetails?.aws?.account_id);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isFetchingDetails, isSuccessDetails]);
  });
```

**third solution: do nothing:**

But maybe, silencing warnings will just make us forget we're using wrong patterns, and what if we want to fix them in the future? Maybe it's good to have them listed somewhere so that we can address them once and for all?


**conclusion**

My own opinion is that we should address the `useEffect` by following the guidelines on the React page. But I'll let @regexowl and @lucasgarfield give their opinions too here.

Lemme know if you have any questions.

Reply:
with your suggestion change without ussing useEffect const [previousFetchingDetails, setPreviousFetchingDetails] = useState(isFetchingDetails);
  const [previousSuccessDetails, setPreviousSuccessDetails] = useState(isSuccessDetails);
  if (isFetchingDetails !== previousFetchingDetails || isSuccessDetails !== previousSuccessDetails) {
    setPreviousFetchingDetails(isFetchingDetails);
    setPreviousSuccessDetails(isSuccessDetails);
    if (!(isFetchingDetails || !isSuccessDetails)) {
      change('aws-associated-account-id', sourceDetails?.aws?.account_id);
    }
  }
  I get this error   ● Step Upload to AWS › validation works

    Expected test not to call console.error().

    If the error is expected, test for it explicitly by mocking it out using jest.spyOn(console, 'error').mockImplementation() and test that the warning occurs.

    Warning: Cannot update a component (`TextField`) while rendering a different component (`AWSSourcesSelect`). To locate the bad setState() call inside `AWSSourcesSelect`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render

      20 |
      21 | export const AWSSourcesSelect = ({
    > 22 |   label,
         |   ^
      23 |   isRequired,
      24 |   className,
      25 |   ...props

Reply:
In general I agree that it’s a good idea to work on removing unneeded useEffect() hooks. But I think we ought to do that in a follow-up PR, and focus on the low hanging fruit in this PR… in this specific case, we can fix the linter warning by simply adding the missing dependencies to the dependencies array as suggested by the linter (just add `change` and `sourceDetails` to the array):

![image](https://github.com/RedHatInsights/image-builder-frontend/assets/95542540/44552f8f-8bb6-45e3-a1f3-da69ad3cdcf8)

Having said that, I think that this is actually one of those edge cases where we do need the useEffect().

If I try first solution proposed, I get the same warning as Michal:

```Warning: Cannot update a component (`TextField`) while rendering a different component (`AWSSourcesSelect`).```

I did some experiments and I believe the problem is the call to change() with `aws-associated-account-id`, as it causes a re-render of <TextField> responsible for rendering the Associated Account ID  while rendering <AWSSourcesSelect>. Screenshot for reference:

![image](https://github.com/RedHatInsights/image-builder-frontend/assets/95542540/eefc19d5-b9b2-4067-828f-6d7089fb60f6)

That warning was introduced in React v16, and they added some notes about it:
https://legacy.reactjs.org/blog/2020/02/26/react-v16.13.0.html#warnings-for-some-updates-during-render

>This warning will help you find application bugs caused by unintentional state changes. In the rare case that you intentionally want to change the state of another component as a result of rendering, you can wrap the setState call into useEffect.

From the “You might not need an Effect” article from the React docs:

>Don’t rush to add Effects to your components. Keep in mind that Effects are typically used to “step out” of your React code and synchronize with some external system. This includes browser APIs, third-party widgets, network, and so on. If your Effect only adjusts some state based on other state, you might not need an Effect.

In this case, the ‘external system’ we need to synchronize with is the state managed by Data Driven Forms. So my interpretation is that it is appropriate here – and within the React guidelines – for us to use the useEffect hook in this case.

Reply:
as we discussed in the group channel, we agreed to add those dependencies
and in the future maybe we will refector the code and remove the useEffect hook were ever we can